### PR TITLE
Use React on the company edit form

### PR DESCRIPTION
**Pull request by rafenden**
## Description of change

Use React on the edit company form, the old form will be removed in a following PR.

## Test instructions

You should be able to edit company as before.
 
### Before

localhost_3000_companies_e667efd0-00c0-e511-88b6-e4115bead28a_edit-old

### After 

localhost_3000_companies_e667efd0-00c0-e511-88b6-e4115bead28a_edit

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied. 
https://github.com/uktrade/data-hub-frontend/blob/develop/docs/Code%20review%20guidelines.md"

- [ ] Has the branch been rebased to develop?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/templates/_layouts/template.njk [hunk 1]
```diff
@@ -126,7 +126,8 @@
 {% block footer %}{% endblock %}
 
 {% block bodyEnd %}
-  <div id="react-app">
+  {% set props = { csrfToken: csrfToken } %}
+  <div id="react-app" data-props="{{ props | dump }}">
```
Comment (yozzo) [inline src/templates/_layouts/template.njk]: dump still needed?

File: src/apps/companies/apps/edit-company/__test__/controllers.test.js [hunk 2]
```diff
@@ -0,0 +1,258 @@
+const config = require('~/config')
```
Comment (yozzo) [inline src/apps/companies/apps/edit-company/__test__/controllers.test.js]: are we still using wavy, or is it only for tests?

File: src/apps/companies/apps/edit-company/client/EditCompanyForm.jsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+  async function onSubmit (values) {
+    await axios.post(`/companies/${companyDetails.id}/edit`, values, {
+      params: { _csrf: csrfToken },
+    })
+    return returnUrl
+  }
+
+  // TODO: Support nested form values to avoid transformation
+  return (
+    <Form onSubmit={onSubmit} initialValues={companyDetails}>
```
Comment (yozzo) [inline src/apps/companies/apps/edit-company/client/EditCompanyForm.jsx]: I think we could break this form in smaller bits:
 - CompanyDetails 
 - CompanyAddress
 - UkBasedCompany
 - OneListCompany

File: src/apps/companies/apps/edit-company/controllers.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+const { get } = require('lodash')
+
+const config = require('../../../../../config')
+const { transformCompanyToForm, transformFormToCompany } = require('./transformers')
+const { updateCompany } = require('../../repos')
+const { getHeadquarterOptions } = require('./repos')
+const { getOptions } = require('../../../../lib/options')
+
+// Foreign companies can have a UK branch.
+const UK_BRANCH_OF_FOREIGN_COMPANY_ID = 'b0730fc6-fcce-4071-bdab-ba8de4f4fc98'
```
Comment (yozzo) [inline src/apps/companies/apps/edit-company/controllers.js]: I'm assuming this is an universal identifier that applies to all foreign companies that have a UK branch?

File: src/apps/companies/router.js [hunk 5]
```diff
@@ -76,15 +78,18 @@ router.get('/export',
...
+router.use('/:companyId/edit', editCompanyFormRouter)
+
 router.use('/:companyId/lists/create', addCreateListFormRouter)
 
 router
   .route('/:companyId/exports/edit')
   .get(populateExportForm, renderExportEdit)
   .post(populateExportForm, handleEditFormPost, renderExportEdit)
 
+// TODO: Remove in a separate PR
```
Comment (yozzo) [inline src/apps/companies/router.js]: shall we?

File: src/apps/companies/views/business-details.njk [hunk 6]
```diff
@@ -1,22 +1,25 @@
...
-            {
-              value: 'Registered',
-              label: 'Type',
-              type: 'badge'
-            }
-          ],
-          modifier: 'stacked',
-          itemModifier: 'stacked'
-        })
+      MetaList({
```
Comment (yozzo) [inline src/apps/companies/views/business-details.njk]: this should be indented one tab

File: src/apps/companies/apps/add-company/__test__/controllers.test.js [hunk 7]
```diff
@@ -90,8 +90,6 @@ describe('Add company form controllers', () => {
 
         expect(this.middlewareParameters.resMock.render).to.be.calledOnceWithExactly(expectedTemplate, {
```
Comment (ian-leggett) [inline src/apps/companies/apps/add-company/__test__/controllers.test.js]: Could we remove `this`

File: src/apps/companies/apps/edit-company/__test__/repos.test.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+const token = 'abcd'
+
+describe('Edit company form repos', () => {
+  describe('#getHeadquarterOptions', () => {
+    beforeEach(async () => {
+      nock(config.apiRoot)
+        .get('/v4/metadata/headquarter-type')
+        .reply(200, headquarterTypeFixture)
+
+      this.actual = await getHeadquarterOptions(token)
```
Comment (ian-leggett) [inline src/apps/companies/apps/edit-company/__test__/repos.test.js]: Can we remove `this`

File: src/apps/companies/apps/edit-company/controllers.js (hunk 4 above)
Comment (ian-leggett) [reply src/apps/companies/apps/edit-company/controllers.js]: Should this be in a config?

**Comment by paulgain** [comment]
When defining a company website URL and saving it everything works as expected, however, if you then remove the same website URL and save this fails as the URL is still present.

**Comment by paulgain** [comment]
Another issue is when editing a field at the bottom of the form you don't see the loading indicator, the user has no visual aid as to what is happening while saving unless they scroll to the top of the form bringing it in to view. I can see this being a problem with all large forms going forward. I realise it's work in progress but we should have a loading indicator that sits on top of the entire web app which is centrally positioned. Axios has the ability to intercept HTTP calls (like all HTTP clients) allowing us to add and remove a loading indicator.

File: src/apps/companies/apps/add-company/__test__/controllers.test.js (hunk 7 above)
Comment (rafenden) [reply src/apps/companies/apps/add-company/__test__/controllers.test.js]: Good point, `this` removed!

File: src/apps/companies/apps/edit-company/client/EditCompanyForm.jsx (hunk 3 above)
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/EditCompanyForm.jsx]: I'm not sure about this, it's hard to split the form like that as UK company could also be on the one list and be matched with DNB record.

File: src/apps/companies/router.js (hunk 5 above)
Comment (rafenden) [reply src/apps/companies/router.js]: Yes, in a separate PR. Want to keep this one short.

File: src/apps/companies/views/business-details.njk (hunk 6 above)
Comment (rafenden) [reply src/apps/companies/views/business-details.njk]: Indentation added.

File: src/templates/_layouts/template.njk (hunk 1 above)
Comment (rafenden) [reply src/templates/_layouts/template.njk]: Yes, as this has to be printed as a JSON string.

**Comment by rafenden** [comment]
> Another issue is when editing a field at the bottom of the form you don't see the loading indicator, the user has no visual aid as to what is happening while saving unless they scroll to the top of the form bringing it in to view. I can see this being a problem with all large forms going forward. I realise it's work in progress but we should have a loading indicator that sits on top of the entire web app which is centrally positioned. Axios has the ability to intercept HTTP calls (like all HTTP clients) allowing us to add and remove a loading indicator.

This is an issue related to the core forms functionality and should be addressed separately.

File: src/apps/companies/apps/edit-company/__test__/repos.test.js (hunk 8 above)
Comment (rafenden) [reply src/apps/companies/apps/edit-company/__test__/repos.test.js]: Done!

**Pull request merged**